synapse/utils/stormcov/plugin.py

```diff
@@ -0,0 +1,182 @@
+        for (dirpath, dirnames, filenames) in os.walk(dirn):
+            for filename in filenames:
+                if regex.search(rx, filename):
+                    path = os.path.join(dirpath, filename)
+                    with open(path, "r") as f:
+                        source = f.read()
+                        tree = self.parser.parse(source)
+                        guid = s_common.guid(str(tree))
+                        self.guid_map[guid] = os.path.abspath(path)
+
+    def file_tracer(self, filename):
+        if filename.endswith('synapse/lib/ast.py'):
+            return self
+        if filename.endswith('synapse/lib/stormctrl.py'):
+            return StormCtrlTracer(self)
+        return None
+
+    def file_reporter(self, filename):
+        return StormReporter(filename, self.parser)
+
+    def find_executable_files(self, src_dir):
+        rx = r"^[^#~!$@%^&*()+=,]+\.(" + "|".join(self.extensions) + r")$"
+        for (dirpath, dirnames, filenames) in os.walk(src_dir):
+            for filename in filenames:
+                if regex.search(rx, filename):
+                    path = os.path.join(dirpath, filename)
+                    yield path
+
+    def has_dynamic_source_filename(self):
+        return True
+
+    def _parent_node(self, node):
+        if not hasattr(node, 'parent'):
+            return node
+        return self._parent_node(node.parent)
+
+    PARSE_METHODS = {"compute", "once"}
+
+    def dynamic_source_filename(self, filename, frame, force=False):
+        if frame.f_code.co_name not in self.PARSE_METHODS and not force:
+            return None
+
+        node = frame.f_locals.get('self')
+        pnode = self._parent_node(node)
+        filename = self.node_map.get(id(pnode), s_common.novalu)
+        if filename is not s_common.novalu:
+            return filename
+
+        if not pnode.__class__.__name__ == 'Query':
+            self.node_map[id(pnode)] = None
+            return
+
+        filename = self.text_map.get(pnode.text)
+        if filename:
+            return filename
+
+        tree = self.parser.parse(pnode.text)
+        guid = s_common.guid(str(tree))
+        filename = self.guid_map.get(guid)
+
+        self.node_map[id(pnode)] = filename
+        self.text_map[pnode.text] = filename
+        return filename
+
+    def line_number_range(self, frame):
+        return frame.f_locals.get('self').lines
+
+class StormCtrlTracer(coverage.FileTracer):
+    def __init__(self, parent):
+        self.parent = parent
+
+    def has_dynamic_source_filename(self):
+        return True
+
+    def dynamic_source_filename(self, filename, frame):
+        if frame.f_code.co_name != '__init__':
+            return None
+        return self.parent.dynamic_source_filename(None, frame.f_back, force=True)
+
+    def line_number_range(self, frame):
+        return self.parent.line_number_range(frame.f_back)
+
+SHOW_PARSING = False
+
+TOKENS = [
```

Comment:
Does this map to another token list we have ??? If so, can we reuse it to avoid multi-master problems?

Reply:
Not really unfortunately